[2단계 - 웹 자동차 경주] 모디(전제희) 미션 제출합니다.

jaehee329 opened the pull request:
안녕하세요 또링~🙌  

2단계 기능들은 구현이 되었고 궁금한 점이 있어 리뷰받고자 합니다!

- - - -

요구사항을 해결한 방향은 아래와 같습니다.
1. 기존의 메인 서비스 하위의 전체 조회를 하는 `FindRecordService`를 만들어주어 전체 조회의 책임을 맡도록 했습니다.
2. N+1 문제를 피하기 위해 세 개의 테이블을 Join해서 가져오는 기능을 DAO에 추가했습니다.
3. 예외 발생 시 View까지 노출되지는 않고, HttpRequest 400이 반환되는 정도로만 구현했습니다.

- - - -
2번을 구현하며 궁금증이 많이 생겼는데요,  
우선 Join을 수행하는 Dao는 어떤 Entity에 대응되는 DAO에서 해줘야 할까요?  

1단계까진 DAO는 Entity와 1:1 대응이 되어 어떤 Entity의 DAO에서 `findById()`와 같은 기능을 수행하면 해당 DAO에 대응되는 Entity를 반환하면 되었습니다. Join을 시작하는 원 테이블에 대응되는 DAO에서 Join 쿼리를 보내야 한다면 그 메서드의 반환형은 어떤 새로운 자료형으로 해야 할까요? 또 Entity라 보기 어려운데 네이밍은 어떻게 해야 할까요? DAO 클래스 각각을 나눌 전략에 대한 조언을 듣고 싶습니다.  
현재는 임시로 별도의 DAO(`FindAllRecordsDao`)와 반환을 위한 자료형(`Result`)을 만들어두었습니다. 조언을 받고 수정하겠습니다!  

스프링보다 DB쪽에 너무 시간을 많이 쏟은 듯 하고 같은 코드만 오래 보다 보니 사소하게 놓친 부분들이 꽤 있을 것 같네요..  
스프링 쪽에서 제대로 알고 있나 물어봐주시는 질문도 너무 감사하고 사소한 부분 다 좋습니다~~~! 잘 부탁드립니다🙇🙇‍♀️🙇🏻

jaehee329 committed:
refactor: DAO를 통해 매핑되는 엔티티 필드의 자료형 변경

ID를 long -> Long으로 받아 생성 시점엔 0 -> null을 넣도록 변경
Timestamp를 String -> java.sql.Timestamp를 받도록 변경

jaehee329 committed:
refactor: 테이블 이름을 단수 명사로 통일

users 관련 entity, dao에서 users => user
game_users_position => position
game_win_users => winner

jaehee329 committed:
refactor: Entity 클래스들의 생성자 수정

일부 값만 입력하는 생성자들 일괄 수정

jaehee329 committed:
feat: 자동차 경주 게임의 모든 플레이 이력 반환 로직 구현

jaehee329 committed:
refactor: ResponseDto가 Car이 아닌 CarDto를 의존하도록 변경

jaehee329 committed:
feat: console app의 출력을 web app과 동일하게 변경

jaehee329 committed:
refactor: findAll의 쿼리를 join을 통한 단일 쿼리로 변경

jaehee329 committed:
feat: 예외 발생 시 HTTP 400이 반환되도록 설정

발생한 예외의 이름을 Body에 담아 보내도록 설정

jaehee329 committed:
refactor: 콘솔 앱이 메인 서비스를 통해 작동하도록 수정

kang-hyungu commented on src/main/java/racingcar/ConsoleRacingCarApplication.java:
```diff
@@ -0,0 +1,28 @@
+import racingcar.service.PlayRacingCarService;
+import racingcar.service.SaveRacingCarResultService;
+
+public class ConsoleRacingCarApplication {
```
콘솔 게임인데 DAO, 서비스, 컨트롤러가 필요한가요? 너무 복잡한 구조 아닌가요?

kang-hyungu commented on src/main/java/racingcar/controller/ConsoleRacingCarController.java:
```diff
@@ -0,0 +1,31 @@
+import racingcar.view.InputView;
+import racingcar.view.OutputView;
+
+public class ConsoleRacingCarController {
```
이게 왜 컨트롤러죠?

kang-hyungu commented on src/main/java/racingcar/controller/ConsoleRacingCarController.java:
```diff
@@ -0,0 +1,31 @@
+        try {
+            String names = InputView.readCarNames();
+            int attempt = InputView.readAttemptNumber();
+            RequestDto requestDto = new RequestDto(names, attempt);
```
1. 왜 Request라는 이름을 붙였죠?
2. Dto가 필요한가요?

kang-hyungu commented on src/main/java/racingcar/controller/ConsoleRacingCarController.java:
```diff
@@ -0,0 +1,31 @@
+            String names = InputView.readCarNames();
+            int attempt = InputView.readAttemptNumber();
+            RequestDto requestDto = new RequestDto(names, attempt);
+            ResponseDto responseDto = mainRacingCarService.raceCar(requestDto);
```
서비스가 필요한가요?

kang-hyungu commented on src/main/java/racingcar/controller/ExceptionAdvice.java:
```diff
@@ -0,0 +1,20 @@
+
+    @ExceptionHandler(DataAccessException.class)
+    public ResponseEntity<String> handleDataAccessException() {
+        return ResponseEntity.badRequest().body("DataAccessException");
```
response body에 DataAccessException이라는 메시지를 반환하는 이유가 뭔가요?
사용자에게 저 메시지가 그대로 노출될텐데, 사용자가 저 메시지를 보고 어떤 액션을 취할 수 있나요? 의미 있는 메시지인가요?

jaehee329 replied:
아직 로깅과 예외 처리를 어떤 식으로 할 지 자세히 공부하지 않아서 간단히만 처리했습니다!
이번에 수정하여 로깅을 통해 서버에도 기록을 남기고 응답에 예외 메시지를 자세히 던지도록 변경했습니다
예외에 대한 자세한 안내는 Alert 창을 띄우는 등으로 처리를 할 수 있을 듯 한데 이런 내용은 타임리프를 조금 사용하면 될 것이나 당장은 다른 것들에 집중했습니다

kang-hyungu commented on src/main/java/racingcar/controller/WebRacingCarController.java:
```diff
@@ -1,25 +1,37 @@
 
-@RestController
-public class RacingCarWebController {
+@Controller
```
`@RestController`를 `@Controller`로 바꾼 이유가 있을까요?

jaehee329 replied:
다시 고민해보니 `@RestController`를 사용하는게 맞네요. 수정했습니다!

kang-hyungu commented on src/main/java/racingcar/controller/WebRacingCarController.java:
```diff
@@ -1,25 +1,37 @@
 
-    @PostMapping("/plays")
+    @GetMapping
+    public ResponseEntity<List<ResponseDto>> findAllRecords() {
```
ResponseDto라는 이름으로 무슨 값을 담고 있는지 알 수가 없네요.
범용적으로 쓸 수 있는 이름을 붙이지 마세요. 어떤 용도인지 이름으로 파악할 수 있게 바꿔보세요.

jaehee329 replied:
RacingCarGameRequest, RacingCarGameResponse로 변경했습니다!

kang-hyungu commented on src/main/java/racingcar/controller/WebRacingCarController.java:
```diff
@@ -1,25 +1,37 @@
+
+    @PostMapping
+    @ResponseBody
     public ResponseEntity<ResponseDto> playRacingCar(@RequestBody RequestDto requestDto) {
```
RequestDto도 마찬가지. 네이밍에 대해 고민해보세요.

kang-hyungu commented on src/main/java/racingcar/dto/ResponseDto.java:
```diff
@@ -1,27 +1,26 @@
 
-    public ResponseDto(final List<String> winners, final List<Car> racingCars) {
+    public ResponseDto(final List<String> winners, final List<CarDto> racingCars) {
         this(String.join(",", winners), racingCars);
```
dto에서 문자열 join 처리를 하고 있군요. 이게 dto 맞을까요?

jaehee329 replied:
말씀을 듣고 보니 도메인 로직을 가지지 않는 것이 DTO이므로 이에 맞게 없애는 것이 맞다는 생각이 드네요.

대신 DTO에서 자신이 변환해야 할 객체를 아는 것은 어떤가요? 현재는 정적 팩토리로 변환할 객체를 넣어 알아서 생성하도록 했습니다.
전에 리뷰어와 얘기하며 DTO는 도메인 객체를 알아도 괜찮으며, 변환된 데이터의 전송을 담당하므로 객체 변환에 대한 로직정도는 가져도 괜찮다는 의견에 공감을 했었습니다.

kang-hyungu commented on src/main/java/racingcar/repository/RacingCarRepository.java:
```diff
@@ -0,0 +1,14 @@
+
+public interface RacingCarRepository {
+
+    List<Record> findAllRecords();
```
```suggestion
    List<Record> findAll();
```

리턴 타입이 Record인데 메서드명에 또 쓸 필요가 있나요?

kang-hyungu commented on src/main/java/racingcar/repository/RacingCarRepository.java:
```diff
@@ -0,0 +1,14 @@
+
+    List<Record> findAllRecords();
+
+    void saveRacingCarResult(final RacingCarResult racingCarResult);
```
```suggestion
    void save(final RacingCarResult racingCarResult);
```

kang-hyungu commented on src/main/java/racingcar/repository/dao/ConsoleFindAllRecordDao.java:
```diff
@@ -0,0 +1,14 @@
+
+import racingcar.repository.entity.Record;
+
+public class ConsoleFindAllRecordDao implements FindAllRecordsDao {
```
콘솔에 이게 왜 필요한가요????

kang-hyungu commented on src/main/java/racingcar/repository/dao/ConsoleWinnerDao.java:
```diff
@@ -0,0 +1,19 @@
+
+import racingcar.repository.entity.WinnerEntity;
+
+public class ConsoleWinnerDao implements WinnerDao {
```
Console로 시작하는 Dao가 필요한가요??

kang-hyungu commented on src/main/java/racingcar/repository/RacingCarRepository.java:
```diff
@@ -0,0 +1,14 @@
+import racingcar.repository.entity.Record;
+import racingcar.service.RacingCarResult;
+
+public interface RacingCarRepository {
```
이 인터페이스는 안 쓰는 것 같으니 정리하시죠.

kang-hyungu commented on src/main/java/racingcar/repository/entity/GameEntity.java:
```diff
@@ -1,23 +1,23 @@
+    private final Timestamp lastModifiedTime;
 
-    public GameEntity(final long id, final int trialCount, final String lastModifiedTime) {
+    public GameEntity(final Long id, final int trialCount, final Timestamp lastModifiedTime) {
```
Timestamp 보다 LocalDateTime을 쓰시기 바랍니다.
Entity로 정의했으니 Timestamp를 쓸 수도 있겠지만 다른 자바 객체에서 Timestamp 사용하게 되면 의도치 않게 도메인 객체가 sql 관련 객체를 의존하게 됩니다.

jaehee329 replied:
감사합니다, Date로 받고 이를 LocalDateTime으로 변환하는 방법이 가장 적절하겠네요!

kang-hyungu commented on src/main/java/racingcar/repository/entity/Record.java:
```diff
@@ -0,0 +1,26 @@
+package racingcar.repository.entity;
+
+public class Record {
```
엔티티 맞나요?

jaehee329 replied:
음.. 위의 질문에도 올리긴 했는데요,  
join을 어떤 DAO에서 할 지, join문으로 가져온 결과를 매핑할 객체가 필요한데 이 객체는 엔티티가 아닌데 무엇이라고 하고 어디에 둘 지가 애매하더라구요, 힌트 주시면 더 고민해보겠습니다!

kang-hyungu commented on src/main/java/racingcar/service/FindRecordService.java:
```diff
@@ -0,0 +1,50 @@
+        return new RacingCarResult(winners, cars);
+    }
+
+    private List<String> findWinners(final List<Record> records) {
```
레벨1에서 배운 객체지향 어디간거죠...
1. Winners는 왜 문자열로 관리하나요? 문자열 타입만 보고 Winners인지 판단이 될까요?
2. 왜 메서드 이름과 다른 일을 하나요.
3. 게임 로직을 서비스에서 처리하지 마세요. 도메인 객체로 옮기세요.

jaehee329 replied:
여기서는 조회 결과에 누락된 winner 정보를 서비스에서 재구성하려고 했었습니다.  
위치를 기반으로 승자를 찾는 게 도메인 로직이니 도메인에 넣는 게 맞는 것 같으면서도,  
DAO의 조회 결과를 Controller에 필요한 형태로 가공하는 것은 서비스의 역할이니 여기서 하는것도 맞아 보여서 이렇게 했습니다.  

현재는 join 쿼리 자체를 수정해서 winner 정보 또한 가져오도록 해 이런 부분이 없도록 바꿨습니다!

kang-hyungu commented on src/main/java/racingcar/service/FindRecordService.java:
```diff
@@ -0,0 +1,50 @@
+        this.findAllRecordsDao = findAllRecordsDao;
+    }
+
+    public List<RacingCarResult> findAllRecords() {
```
왜 이렇게 작성하셨나요?
요 메서드에 대해 이야기 좀 들어봐야겠네요.

kang-hyungu commented on src/main/java/racingcar/service/MainRacingCarService.java:
```diff
@@ -1,31 +1,43 @@
 import racingcar.dto.ResponseDto;
 
 @Service
 public class MainRacingCarService {
```
이건 무슨 클래스죠??? 다른 팀원이 메인 레이싱 카라는 이름을 보면 물어볼 수 밖에 없지 않을까요?

kang-hyungu commented on src/main/java/racingcar/service/RacingCarResult.java:
```diff
@@ -7,14 +7,18 @@ public class RacingCarResult {
+    private final Integer attempt;
 
-    public RacingCarResult(final List<String> winners, final List<Car> cars, final int attempt) {
+    public RacingCarResult(final List<String> winners, final List<Car> cars, final Integer attempt) {
```
이 클래스는 왜 서비스로 두셨나요?

jaehee329 replied:
서비스는 아니고 서비스 간 데이터 전달용 객체가 필요했는데 DTO라 명명할지 확신이 안 서서요,  
서비스 하위에 DTO 패키지를 만들어 옮겼습니다!

kang-hyungu commented on src/main/java/racingcar/ConsoleRacingCarApplication.java:
```diff
@@ -0,0 +1,28 @@
+
+public class ConsoleRacingCarApplication {
+
+    public static void main(String[] args) {
```
1. 메인에서 객체 주입하도록 만들었는데 RandomNumberGenerator는 왜 여기서 생성하지 않나요?
2. RandomNumberGenerator 패키지가 util이라고 볼 수 있나요???

jaehee329 replied:
말씀대로 NumberGenerator를 메인에서 주입하는 것이 맞는데 제가 누락했네요!  
RandomNumberGenerator는 도메인 객체로 옮기고 웹 앱에서는 Bean으로 등록하여 사용할 수 있도록 변경했습니다~!

kang-hyungu reviewed (changes requested):
모디 수정해보시고 리뷰 재요청 주세요.
리뷰 재요청하기 전에 내일 코치룸으로 오시기 바랍니다.
코드를 보니 물어볼게 많네요.

jaehee329 committed:
refactor: 예외 발생 시 로그 출력 및 사용자에게 예외 메시지 출력하도록 변경

jaehee329 committed:
fix: 웹 컨트롤러가 @RestController를 사용하도록 변경

jaehee329 committed:
refactor: Controller, Service 사이의 DTO 변경

DTO 네이밍 변경 및 도메인 로직의 변환 책임 부여

jaehee329 committed:
refactor: DAO 관련 메서드 네이밍 변경

jaehee329 committed:
refactor: DAO 가 SimpleJdbcInsert를 사용하도록 변경

jaehee329 committed:
refactor: 쓰이지 않는 DAO 메서드 제거

jaehee329 committed:
refactor: NumberGenerator를 Bean으로 등록하도록 변경

jaehee329 committed:
refactor: find record 관련 서비스 및 dao 변경

변경 전: dao를 통해 가져온 이름, 포지션으로 서비스에서 승자 계산
변경 후: winner 테이블까지 join하여 완결된 결과를 가져오도록 함

jaehee329 committed:
refactor: DAO 변경에 의한 불가능한 테스트 삭제

jaehee329 committed:
refactor: controller - service 간 전달 객체 수정

controller에서 view로 전달되는 request, response 인스턴스 생성, 분해

jaehee329 committed:
refactor: Service의 메서드, 변수 네이밍 변경

jaehee329 committed:
test: DAO에 대한 테스트 추가

jaehee329 committed:
refactor: 콘솔 게임에서 서비스를 사용하지 않도록 변경

console 관련 dao 삭제

jaehee329 commented:
안녕하세요~  
구구가 말씀하신 대로 콘솔에서 서비스 및 DAO를 사용하지 않도록 수정했습니다.
- - - -
[이 질문](https://github.com/woowacourse/jwp-racingcar/pull/132#discussion_r1172028894)에 대해서도 오래 고민했는데요, 구구가 생각하신 설계가 무엇이었는지는 아직도 잘 모르겠습니다.  
콘솔 어플리케이션에서는 기존 코드 그대로 순수 자바 어플리케이션으로 돌아가기 때문에 스프링 웹앱과 달리 전체 프로그램의 흐름을 제어하는 것이 컨트롤러의 역할이라고 생각했습니다. 여태 1단계에서 이런 방향으로 구현했었구요.  

질문이 스프링 프레임워크에서의 컨트롤러와는 구조적인 역할이 너무 다르다는 것이었다면,  
따로 말씀해주신 것처럼 이 역할을 모두 psvm이 존재하는 메인 애플리케이션 클래스에서 수행해도 안될 것은 없다고 생각합니다.  

만약 질문이 콘솔 컨트롤러 클래스를 유지하면서 제가 책임 분리를 잘못한 것을 지적한 것이라면,  
어떤 책임을 메인 앱 클래스와 콘솔 컨트롤러로 분리하는게 맞다고 생각하시는지 힌트를 부탁드려도 될까요?  
예외 시 재시작이나 View의 접근해야 함으로 인해서 하나의 클래스의 처리가 나을 것 같아서요.  
이 또한 아니었다면 어떤 의미였는지 설명해주신다면 다시 고민해보겠습니다.  
콘솔 앱의 예외 처리 및 재시작, 출력 등은 이번 미션에서 중요도가 낮아 간단히만 처리했는데 피드백 후 필요하다면 보완하겠습니다

- - - -
말씀해주신 작은 부분들이 많았었는데 가장 주요한 건 도메인 객체와 엔티티들의 매칭이 안 된다는 것이었던 걸로 기억하는데요,  
남은 시간 동안 테이블을 재설계하며 관련 사항들을 바꿔보겠습니다  

감사합니다~

jaehee329 committed:
refactor: 도메인과 일치하도록 테이블 명 변경

ddl: users -> player, position -> car
domain: Name -> Player

jaehee329 committed:
fix: 게임 정보 업데이트 시점이 아닌 실행 시점을 저장하도록 변경

jaehee329 committed:
refactor: 테이블 join 결과를 dto 객체에 매핑하도록 변경

repository.entity.record => repository.dto.result

jaehee329 committed:
refactor: 사용자가 존재하는지 판별하는 로직 변경

기존: try~catch로 이름으로 찾았을 때 예외가 발생하는지로 존재 판별
변경 후: exist문으로 존재하는지 검사하는 별도 쿼리 사용

jaehee329 committed:
test: findAllResultsDao에 대한 test 작성

jaehee329 commented on src/main/java/racingcar/repository/dao/JdbcFindAllResultsDao.java:
```diff
@@ -0,0 +1,36 @@
+            + "LEFT OUTER JOIN game AS g ON c.game_id = g.id "
+            + "LEFT OUTER JOIN player AS p ON c.player_id = p.id "
+            + "LEFT OUTER JOIN winner AS w ON (g.id = w.game_id AND p.id = w.player_id)";
+        return jdbcTemplate.query(sql, actorRowMapper);
```
정규화의 단점을 전체 결과를 join 하며 많이 느꼈습니다.
하나의 결과로 매핑하며 게임 당 차의 위치와 이름, 그리고 승리 여부를 boolean으로 조회하도록 했는데요,  
구구는 조회 문에 IS NOT NULL 같은 별도의 로직을 넣는 것을 지양하라고 말씀하셨습니다.  
이 이유를 생각해 보았을 때
1. 쿼리만 보고 어떤 기능인지 알기 쉽지 않다.
2. DB의 커넥션이 오래 유지된다.

정도가 떠올랐는데요, 2번의 경우는 100만번의 쿼리를 IS NOT NULL 여부에 따라 테스트해본 결과 약 5% 시간적 차이가 있음을 확인했습니다.  
탐색한 차가 해당 게임에서 승리했는지 여부는 결국 DB아니면 서비스에서 해주어야 하는데요,  
DB에서 하면 안 되는 이유가 여러 웹서버 쓰레드의 요청을 처리하는 하나의 DB에 커넥션이 오래 맺어지면 성능이 치명적인지, (수정이 아닌 조회 쿼리임에도 불구하고) 

만약 그렇다면 IS NOT NULL 없이 winner의 id까지만 받아와서 서비스에서 NULL 여부를 한 번 더 체크해줘야 하는데 이렇게 수정하는게 맞을지 궁금합니다.

jnsorn replied:
좋은 고민을 하셨네요! 👍 

> DB에서 하면 안 되는 이유가 여러 웹서버 쓰레드의 요청을 처리하는 하나의 DB에 커넥션이 오래 맺어지면 성능이 치명적인지, (수정이 아닌 조회 쿼리임에도 불구하고)

DB Connection에 대한 학습은 이전 단계에서 조금씩 하셨을 것 같은데요, 알고 계신대로 커넥션은 pool이 고갈되어 시스템 전체에 영향을 끼치지 않도록 최대한 짧게 유지하는게 중요해요. 그것이 정녕 조회쿼리일지라도요. (다만 writer가 아닌 read는 별도의 reader용 인스턴스를 사용하기 때문에 조금은 널널하게 사용하는 편입니다.)

>  IS NOT NULL 없이 winner의 id까지만 받아와서 서비스에서 NULL 여부를 한 번 더 체크해줘야 하는데 이렇게 수정하는게 맞을지 궁금합니다.

사실 답은 없지만 저도 이런 방식으로 구현하는걸 선호하는 편이긴해요. 쿼리문에 구체적인 조건이 들어가면, 상황마다 조건이 다른 n개의 쿼리문을 정의해주어야하기 때문에 나중에 코드를 유지보수하기도 힘들어지더라구요. 
성능 측면에서는 데이터의 양이나 쿼리에 따라 인덱스를 설정해주고 쿼리에서 조건을 구성하는게 나을수도 있어서, 무조건 성능때문에 애플리케이션단에서 처리해줘야해!는 아닌 것 같습니다 :) 

이 부분은 모디가 조금 더 고민해보시고 결정하시면 될 것 같아요. 모디는 어떤 이유로 조회문에 별도 로직을 들어가게 구성하셨었나요?

jnsorn commented on src/main/java/racingcar/controller/ExceptionAdvice.java:
```diff
@@ -0,0 +1,25 @@
+    public ResponseEntity<String> handleIllegalArgumentException(IllegalArgumentException e) {
+        log.error("IllegalArgumentException : " + e.getMessage());
+        return ResponseEntity.internalServerError().body(e.getMessage());
+    }
```
사용자의 이름, 시도 횟수등을 잘못 입력했을 경우 `IllegalArgumentException`을 발생시키고 있는데요, 응답코드는 500을 내려주고 있어요. 잘못된 입력에 의해 500 에러를 내려주는 것이 모디가 의도한 동작이 맞을까요?

jaehee329 replied:
badRequest가 더 맞는 동작이겠네요,  
별개로 보통 서버의 문제여도 그렇다고 응답하지 않는 것으로 알고 있는데요, 이런 경우 보통 504를 사용하나요?  
일반적인 선택이 무엇인지 궁금합니다.

jnsorn replied:
badRequest 👍

> 별개로 보통 서버의 문제여도 그렇다고 응답하지 않는 것으로 알고 있는데요, 이런 경우 보통 504를 사용하나요?
> 일반적인 선택이 무엇인지 궁금합니다.

서버의 문제여도 그렇다고 응답하지 않는 것
-> 오! 아뇨. 서버에 문제가 발생하면 클라이언트가 서버의 상황을 알 수 있도록 적절한 응답 코드를 통해 상황을 파악할 수 있도록 해야합니다. 다만 사용자에게 자세한 메세지는 노출시키지 않아요. (보통은 500을 내리는 경우가 많아요) 
각 응답코드에 대해 학습해봐도 좋겠네요!

jnsorn commented on src/main/java/racingcar/controller/ExceptionAdvice.java:
```diff
@@ -0,0 +1,25 @@
+public class ExceptionAdvice {
+    private final Logger log = LoggerFactory.getLogger(this.getClass().getSimpleName());
+
+    @ExceptionHandler(DataAccessException.class)
+    public ResponseEntity<String> handleDataAccessException() {
+        log.error("DataAccessException");
+        return ResponseEntity.internalServerError().body("DataAccessException");
```
`DataAccessException`를 받은 사용자는 어떤 행동을 취할 수 있을까요? 서버 상태를 사용자에게 구체적으로 노출할 필요가 있을까요? '잠시 후 다시 시도해주세요'와 같은 문구를 내려주면 어떨까요?

jnsorn replied:
https://github.com/woowacourse/jwp-racingcar/pull/132#discussion_r1173222008 
이 코멘트를 이제 확인했네요. 현재 예외처리보다 다른 것에 집중하고 싶다면, 이 리뷰는 스킵하셔도 좋습니다. :)

jaehee329 replied:
당장은 View에서 Alert를 띄우는 등의 추가적인 처리를 하진 않아서 임시로 예외를 그대로 노출하도록 하였습니다.  
말씀처럼 DataAcessException의 경우는 서버 상태를 감추고, IllegalArgumentException은 메시지를 출력하도록 수정하였습니다.

jnsorn commented on src/main/java/racingcar/controller/ExceptionAdvice.java:
```diff
@@ -0,0 +1,25 @@
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice
```
`RestControllerAdvice`도 있어요! 😃

jnsorn commented on src/main/java/racingcar/controller/ExceptionAdvice.java:
```diff
@@ -0,0 +1,25 @@
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice
+public class ExceptionAdvice {
```
만약 NPE가 발생한다면 사용자는 어떤 화면을 보게될까요? 이런 예외들에 대해서는 처리해주지 않아도 괜찮을까요?

jaehee329 replied:
현재는 도메인 로직에서 IllegalArgumentException만이 발생해서 이 예외만을 처리하도록 했는데요, 
NPE가 발생할 여지가 있을지는 모르겠지만 그런다면 사용자에게 아무런 응답도 나타나지 않겠네요.  
필요하다면 해주는 것이 맞을 듯 한데 이런 예외가 발생할 상황이 있는데 제가 놓친 것인가요?

jnsorn replied:
네, 예기치 못한 상황에도 사용자는 서버가 정의한 메세지를 확인할 수 있도록 처리해두어야 하는데요, 이런 경우 보통 `Exception.class`를 처리하는 메서드를 만들어 "서버에 문제가 발생했습니다"와 같은 응답을 내리는 편입니다.

jnsorn commented on src/main/java/racingcar/controller/ConsoleRacingCarController.java:
```diff
@@ -0,0 +1,28 @@
+
+public class ConsoleRacingCarController {
+
+    public void run(final NumberGenerator numberGenerator) {
```
콘솔 게임을 도메인을 이용해서 잘 구현해주셨네요 👍

jnsorn commented on src/main/java/racingcar/repository/dao/JdbcCarDao.java:
```diff
@@ -0,0 +1,28 @@
+@Repository
+public class JdbcCarDao implements CarDao {
+
+    private final SimpleJdbcInsert jdbcInsert;
```
SimpleJdbcInsert 👍

jnsorn commented on src/main/java/racingcar/controller/WebRacingCarController.java:
```diff
@@ -0,0 +1,45 @@
+        return ResponseEntity.ok().body(records);
+    }
+
+    @PostMapping
+    public ResponseEntity<RacingCarGameResponse> playRacingCar(@RequestBody RacingCarGameRequest racingCarGameRequest) {
+        final List<String> names = List.of(racingCarGameRequest.getNames().split(","));
+        final int attempt = racingCarGameRequest.getCount();
+        final RacingCarResultDto racingCarResultDto = racingCarService.raceCar(names, attempt);
+        final RacingCarGameResponse racingCarGameResponse = RacingCarGameResponse.from(racingCarResultDto);
+        return ResponseEntity.ok().body(racingCarGameResponse);
+    }
```
게임 결과를 담는 DTO(RacingCarResultDto, RacingCarGameResponse)가 두개 정의되어 있는데요, 각각 어떤 목적의 DTO일까요? 또, 두개로 나누어 사용하신 이유도 궁금해요. 😎

jaehee329 replied:
ResultDto는 서비스와 컨트롤러 간에서, Response는 컨트롤러와 뷰 간에서 사용하였습니다.  
한다면 뷰에서 서비스까지 같은 DTO를 사용할 수는 있었을 것 같은데요,  
뷰에서 넘어오는 요청은 기본적으로 단순한 String 위주이고 서비스에서 사용하는 것은 String을 파싱한 리스트 등으로 차이가 있어서  
레이어 간 의존을 명확히 나눌 겸 분리하였습니다.  
혹시 별도로 고민해보면 하는 이유가 추가적으로 있을까요?

jnsorn replied:
아뇨 단순히 궁금해서 질문드렸습니다. 👍 
사용 이유가 아주 좋은 것 같아요 ㅎㅎ (저도 현업에서 위와 같이 사용하는 편이에요)

jnsorn commented on src/test/java/racingcar/repository/dao/JdbcFindAllResultsDaoTest.java:
```diff
@@ -0,0 +1,73 @@
+        this.findAllResultsDao = new JdbcFindAllResultsDao(dataSource);
+    }
+
+    @Test
+    void findAll_메서드로_모든_게임_결과를_가져온다() {
+
+        final PlayerEntity playerEntity1 = new PlayerEntity("player1");
+        final PlayerEntity playerEntity2 = new PlayerEntity("player2");
+        long player1Id = playerDao.save(playerEntity1);
+        long player2Id = playerDao.save(playerEntity2);
+
+        final GameEntity gameEntity = new GameEntity(1L, 5, LocalDateTime.now());
+        final long gameId = gameDao.save(gameEntity);
+
+        final CarEntity carEntity1 = new CarEntity(gameId, player1Id, 5);
+        final CarEntity carEntity2 = new CarEntity(gameId, player2Id, 5);
+        carDao.save(carEntity1);
+        carDao.save(carEntity2);
+
+        WinnerEntity winner1Entity = new WinnerEntity(gameId, player1Id);
+        WinnerEntity winner2Entity = new WinnerEntity(gameId, player2Id);
+        winnerDao.save(winner1Entity);
+        winnerDao.save(winner2Entity);
+
+
+        List<ResultDto> resultDtos = findAllResultsDao.findAll();
+        SoftAssertions softAssertions = new SoftAssertions();
+        softAssertions.assertThat(resultDtos.size()).isEqualTo(2);
+        softAssertions.assertThat(resultDtos.get(0).getPosition()).isEqualTo(5);
+        softAssertions.assertThat(resultDtos.get(1).getPosition()).isEqualTo(5);
+    }
```
테스트도 꼼꼼하게 잘 짜주셨네요! 👍 
테스트가 길어지는 경우 given, when, then 구절을 나눠 가독성을 높이면 좋습니다. 
```suggestion
    @Test
    void findAll_메서드로_모든_게임_결과를_가져온다() {
       // given
        final PlayerEntity playerEntity1 = new PlayerEntity("player1");
        final PlayerEntity playerEntity2 = new PlayerEntity("player2");
        long player1Id = playerDao.save(playerEntity1);
        long player2Id = playerDao.save(playerEntity2);

        final GameEntity gameEntity = new GameEntity(1L, 5, LocalDateTime.now());
        final long gameId = gameDao.save(gameEntity);

        final CarEntity carEntity1 = new CarEntity(gameId, player1Id, 5);
        final CarEntity carEntity2 = new CarEntity(gameId, player2Id, 5);
        carDao.save(carEntity1);
        carDao.save(carEntity2);

        WinnerEntity winner1Entity = new WinnerEntity(gameId, player1Id);
        WinnerEntity winner2Entity = new WinnerEntity(gameId, player2Id);
        winnerDao.save(winner1Entity);
        winnerDao.save(winner2Entity);

       // when
        List<ResultDto> resultDtos = findAllResultsDao.findAll();
        
        // then 
        SoftAssertions softAssertions = new SoftAssertions();
        softAssertions.assertThat(resultDtos.size()).isEqualTo(2);
        softAssertions.assertThat(resultDtos.get(0).getPosition()).isEqualTo(5);
        softAssertions.assertThat(resultDtos.get(1).getPosition()).isEqualTo(5);
    }
```

jnsorn replied:
현재 carEntity의 position을 5가 아닌 다른 값으로 설정해도 테스트가 통과하는데요, 정상적인 동작일까요?

```
        final CarEntity carEntity1 = new CarEntity(gameId, player1Id, 2);
        final CarEntity carEntity2 = new CarEntity(gameId, player2Id, 2);
```

jaehee329 replied:
soft assertion을 잘못 사용하였네요!  
맞게 동작하도록 수정했습니다

jnsorn reviewed (changes requested):
안녕하세요. 모디! 리뷰가 늦었습니다. 죄송해요 😭
리팩터링을 잘 해주셔서 큰 리뷰는 없는데요, 그래도 같이 얘기해보고 싶은 부분이 있어 코멘트 남겨두었어요. 
질문은 코멘트로 남겨주신것 확인했는데 추가로 질문주실게 있다면 코멘트로 또 남겨주세요!

kang-hyungu commented:
> 안녕하세요~ 구구가 말씀하신 대로 콘솔에서 서비스 및 DAO를 사용하지 않도록 수정했습니다.
> 
> [이 질문](https://github.com/woowacourse/jwp-racingcar/pull/132#discussion_r1172028894)에 대해서도 오래 고민했는데요, 구구가 생각하신 설계가 무엇이었는지는 아직도 잘 모르겠습니다. 콘솔 어플리케이션에서는 기존 코드 그대로 순수 자바 어플리케이션으로 돌아가기 때문에 스프링 웹앱과 달리 전체 프로그램의 흐름을 제어하는 것이 컨트롤러의 역할이라고 생각했습니다. 여태 1단계에서 이런 방향으로 구현했었구요.
> 
> 질문이 스프링 프레임워크에서의 컨트롤러와는 구조적인 역할이 너무 다르다는 것이었다면, 따로 말씀해주신 것처럼 이 역할을 모두 psvm이 존재하는 메인 애플리케이션 클래스에서 수행해도 안될 것은 없다고 생각합니다.
> 
> 만약 질문이 콘솔 컨트롤러 클래스를 유지하면서 제가 책임 분리를 잘못한 것을 지적한 것이라면, 어떤 책임을 메인 앱 클래스와 콘솔 컨트롤러로 분리하는게 맞다고 생각하시는지 힌트를 부탁드려도 될까요? 예외 시 재시작이나 View의 접근해야 함으로 인해서 하나의 클래스의 처리가 나을 것 같아서요. 이 또한 아니었다면 어떤 의미였는지 설명해주신다면 다시 고민해보겠습니다. 콘솔 앱의 예외 처리 및 재시작, 출력 등은 이번 미션에서 중요도가 낮아 간단히만 처리했는데 피드백 후 필요하다면 보완하겠습니다
> 
> 말씀해주신 작은 부분들이 많았었는데 가장 주요한 건 도메인 객체와 엔티티들의 매칭이 안 된다는 것이었던 걸로 기억하는데요, 남은 시간 동안 테이블을 재설계하며 관련 사항들을 바꿔보겠습니다
> 
> 감사합니다~

```java
public class ConsoleRacingCarApplication {

    public static void main(String[] args) {
        final var numberGenerator = new RandomNumberGenerator();
        run(numberGenerator);
    }

    private static void run(final NumberGenerator numberGenerator) {
        try {
            // input
            final var names = List.of(InputView.readCarNames().split(","));
            final var attempt = InputView.readAttemptNumber();

            // process
            final RacingCarGame racingCarGame = new RacingCarGame(names, attempt, numberGenerator);
            racingCarGame.play();

            // output
            OutputView.printResult(racingCarGame.findWinners(), racingCarGame.getCars().getCars());
        } catch (IllegalArgumentException e) {
            OutputView.printError(e);
            run(numberGenerator);
        }
    }
}
```

이렇게 작성하면 콘솔은 컨트롤러가 필요 없겠네요
https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller

구글 번역을 돌려보니 MVC를 창안한 분이 아래와 같은 동기로 만들었다고 하네요.
> MVC의 원래 동기는 모든 객체에 대한 그래픽 인터페이스 생성을 허용하는 것이었습니다
> MVC는 사용자가 크고 복잡한 데이터 세트를 제어하는 ​​문제에 대한 일반적인 솔루션으로 생각되었습니다.

콘솔 자동차 게임이 GUI를 쓰거나 크고 복잡한 데이터 모델을 다루는지 고민해보면 좋을 것 같네요.

jaehee329 committed:
refactor: 예외 발생 시의 처리 수정

jaehee329 committed:
fix: findAllDao 테스트 수정

jaehee329 commented:
감사합니다 구구,  
주변 크루들과도 함께 고민하였는데 항상 불필요하게 메인 앱이 아닌 Controller를 앱의 시작 포인트로 여겼다는 것을 반성했습니다.  
말씀처럼 콘솔에서의 컨트롤러 사용은 불필요한 추상화를 하고 있는 것 같아 제거했습니다

jaehee329 committed:
refactor: 콘솔의 컨트롤러 클래스 삭제

jaehee329 commented:
쿼리문에 조건을 넣은 것에 대해서는,  
아직은 DB의 성능적인 부분을 잘 모르고 현재는 해당 쿼리의 결과가 하나의 결과 객체로 DAO를 통해 가장 편리하게 매핑될 수 있는 방법이어서 그렇게 했습니다.  
두 분의 말씀을 듣고 보니 확실히 조회 시에는 NULL 인 데이터가 있도록 하고 서비스에서 이것을 검사하여 별도의 객체에 매핑시키는 것이 유지보수 측면에서는 더 좋은 방법이겠네요.  
당장은 현행을 유지하고, 이후 유사한 상황들에서 이러한 방법들을 적용해도 괜찮을까요?

jnsorn commented:
그럼요 모디. 여러 방법을 적용해보시면서, 직접 장단점을 느껴보시면 좋을 것 같아요! 👍

jnsorn reviewed (approved):
안녕하세요. 모디! 다음 미션 진행하면서 반영하시느라 고생하셨네요.
리뷰 잘 반영해주신 것 같아요. 💯 (미션 마감기한이 지났기 때문에 별도 리뷰는 드리지 않았습니다)
자동차 경주 미션 수고 많으셨어요. 🚗👍
